runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java

```diff
@@ -119,20 +124,42 @@ public Object getPayload(CorfuRuntime runtime) {
     @Override
     public synchronized void releaseBuffer() {
         if (serializedCache != null) {
-            serializedCache.release();
-            if (serializedCache.refCnt() == 0) {
+            serializedCache.buffer.release();
+            if (serializedCache.buffer.refCnt() == 0) {
                 serializedCache = null;
             }
         }
     }
 
     @Override
-    public synchronized void acquireBuffer() {
+    public synchronized void acquireBuffer(boolean metadata) {
         if (serializedCache == null) {
-            serializedCache = Unpooled.buffer();
-            doSerializeInternal(serializedCache);
+            acquireBufferInternal(metadata);
         } else {
-            serializedCache.retain();
+            if (metadata) {
+                serializedCache.buffer.resetReaderIndex();
+                serializedCache.buffer.writerIndex(serializedCache.metadataOffset);
+                doSerializeMetadataInternal(serializedCache.buffer);
+            }
+            serializedCache.buffer.retain();
+        }
+    }
+
+     public synchronized void updateAcquiredBuffer(boolean metadata) {
+        if (serializedCache == null) {
+            return;
+        }
+        acquireBufferInternal(metadata);
+    }
+
+    private void acquireBufferInternal(boolean metadata) {
```

Comment:
acquireBufferInternsl is really serializing data, making the name misleading. Perhaps something like serializeToNewBuffer()

Reply:
Initially I also wanted to name it serailizeXXX, but we have other methods with this form of name, so to aviod confusion, I changed to use acquireBufferInternal.